src/EVM/SMT.hs

```diff
@@ -405,11 +405,7 @@ assertReads props benv senv = concatMap assertRead allReads
   where
     assertRead :: (Expr EWord, Expr EWord, Expr Buf) -> [Prop]
     assertRead (idx, Lit 32, buf) = [PImpl (PGEq idx (bufLength buf)) (PEq (ReadWord idx buf) (Lit 0))]
-    assertRead (idx, Lit sz, buf) =
-      fmap
-        -- TODO: unsafeInto instead fromIntegral here makes symbolic tests fail
-        (PImpl (PGEq idx (bufLength buf)) . PEq (ReadByte idx buf) . LitByte . fromIntegral)
-        [(0::Int)..unsafeInto sz-1]
+    assertRead (idx, Lit sz, buf) = [PImpl (PGEq (Add idx $ Lit offset) (bufLength buf)) (PEq (ReadByte (Add idx $ Lit offset) buf) (LitByte 0)) | offset <- [(0::W256).. sz-1]]
```

Comment:
I was not sure, but we can probably use the the simplyfing `Expr.add` function here, no?

Reply:
I don't think we should... I think it's best to keep it being used only inside simplify and its subfunctions? I dunno, I am afraid that it will make it harder to reason about code if we don't... but I am no authority on this.

Reply:
But it is being used also outside of that, for example in `EVM.exec1`.
It would just compute the concrete offset value in case `idx` is also concrete.
I was also not sure, but I think it should be OK.

Reply:
Good point. Yeah, let's just do it :) Wanna update the PR?

Reply:
I am always hesitant of side-effects. Of course since it's functional programming, it's not supposed to have side-effects :)

Reply:
OK, I did a commit on top with this, let me know what you think! You can always revert!